lib/spack/spack/concretize.py

```diff
@@ -347,26 +347,46 @@ def concretize_compiler_flags(self, spec):
             # running.
             return True
 
-        compiler_match = lambda other: (
-            spec.compiler == other.compiler and
-            spec.architecture == other.architecture)
+        def compiler_match(other, strict=True):
+            if strict:
+                # Other compiler already matches this compiler.
+                return (spec.compiler == other.compiler and
+                        spec.architecture == other.architecture)
+            else:
+                # Other compiler may match this compiler.
+                if other.compiler and other.compiler.concrete:
+                    return (spec.compiler == other.compiler and
+                            spec.architecture == other.architecture)
+                return True
 
         ret = False
         for flag in spack.spec.FlagMap.valid_compiler_flags():
             if flag not in spec.compiler_flags:
-                spec.compiler_flags[flag] = list()
-            try:
-                nearest = next(p for p in spec.traverse(direction='parents')
-                               if (compiler_match(p) and
-                                   (p is not spec) and
-                                   flag in p.compiler_flags))
-                nearest_flags = set(nearest.compiler_flags.get(flag, []))
-                flags = set(spec.compiler_flags.get(flag, []))
-                if (nearest_flags - flags):
-                    spec.compiler_flags[flag] = list(nearest_flags | flags)
-                    ret = True
-            except StopIteration:
-                pass
+                try:
+                    n = next(p for p in spec.traverse(direction='parents')
+                             if (compiler_match(p) and
+                                 (p is not spec) and
+                                 flag in p.compiler_flags))
+
+                    nearest_flags = set(n.compiler_flags[flag])
+                    flags = set()
+
+                    if (nearest_flags - flags) or n.compiler_flags[flag] == []:
```

Comment:
I would prefer ```or flag in n.compiler_flags``` to ```or n.compiler_flags[flag] == []```. For that matter, if I'm right that ```flag in n.compiler_flags``` implies that n.compiler_flags is a nonempty list, then I think you can strip the "if" altogether and do the next line unconditionally.

Reply:
It being empty means something different than it not being present. Not present means it hasn't been concretized, empty means it has been concretized to empty.

Reply:
* From the logic preceding this statement, we know that `flag in n.compiler_flags == True`
* My understanding is that if `flag in n.compiler_flags`, then it is a list (possibly empty).

If these two statements are correct, then based on how we arrived at this point, `if nearest_flags or n.compiler_flags[flag] == []` is always true, and the `if` isn't required.

On top of that, based on the fact that you moved all of this into the `if` statement, the general logic is "set the flags here if they weren't set before"; but if we are at this point then we know that that self.compiler_flags hasn't been set.